storage/postgreskv/client.go

```diff
@@ -317,3 +317,99 @@ func (client *Client) Iterate(ctx context.Context, opts storage.IterateOptions,
 
 	return fn(ctx, opi)
 }
+
+// CompareAndSwap atomically compares and swaps oldValue with newValue
+func (client *Client) CompareAndSwap(ctx context.Context, key storage.Key, oldValue, newValue storage.Value) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	return client.CompareAndSwapPath(ctx, storage.Key(defaultBucket), key, oldValue, newValue)
+}
+
+// CompareAndSwapPath atomically compares and swaps oldValue with newValue in the given bucket
+func (client *Client) CompareAndSwapPath(ctx context.Context, bucket, key storage.Key, oldValue, newValue storage.Value) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	if key.IsZero() {
+		return storage.ErrEmptyKey.New("")
+	}
+
+	if oldValue == nil && newValue == nil {
+		q := "SELECT metadata FROM pathdata WHERE bucket = $1::BYTEA AND fullpath = $2::BYTEA"
+		row := client.pgConn.QueryRow(q, []byte(bucket), []byte(key))
+		var val []byte
+		err = row.Scan(&val)
+		if err == sql.ErrNoRows {
+			return nil
+		}
+		if err != nil {
+			return Error.Wrap(err)
+		}
+		return storage.ErrValueChanged.New(key.String())
+	}
+
+	if oldValue == nil {
+		q := `INSERT INTO pathdata (bucket, fullpath, metadata) VALUES ($1::BYTEA, $2::BYTEA, $3::BYTEA)`
+		_, err = client.pgConn.Exec(q, []byte(bucket), []byte(key), []byte(newValue))
+		if err, ok := err.(*pq.Error); ok {
+			if err.Code == "23505" { // unique_violation
```

Comment:
This could be

    if err.Code.Name() == "unique_violation" {

I think. That would be slightly safer during future maintenance, but would also be very slightly slower due to a map lookup and a longer string compare. I guess I don't care much which approach you use, but if we use `"23505"` as a constant, I would very much prefer for it to be a named constant.

Reply:
Another approach to this which would avoid inspecting error codes entirely would look like

```sql
    INSERT INTO pathdata (bucket, fullpath, metadata) VALUES ($1::BYTEA, $2::BYTEA, $3::BYTEA)
        ON CONFLICT DO NOTHING
        RETURNING 1
```

You'd run that as though it was a select query. If it returned any rows, then the insert succeeded. If it returned zero rows, then there was a uniqueness violation.

Reply:
You are a wizard! This query actually worked :)